packages/web3-core/src/web3_context.ts

```diff
@@ -220,7 +218,7 @@ export class Web3Context<
 	/**
 	 * Link current context to another context.
 	 */
-	public link<T extends Web3Context<API, RegisteredSubs>>(parentContext: T) {
+	public link<T extends Web3Context>(parentContext: T) {
```

Comment:
`API` should be passed here otherwise the specification of `ChainlinkPluginAPI` is meaningless. The purpose of specifying the `ChainlinkPluginAPI` (as we do with [EthExecutionApi](https://github.com/web3/web3.js/blob/61411f95f03d74ad8e1053c68d6ec0df6ea88417/packages/web3-types/src/apis/eth_execution_api.ts#L215)) is to provide type safety when using the `requestManager` e.g. `API` specifies what `method`s are available to call and their corresponding typed `params` when doing things like so:

![image](https://user-images.githubusercontent.com/20375610/192454412-ff60ece8-67f8-4c33-a865-72287e020e3b.png)

which gives us type safety like so:

![image](https://user-images.githubusercontent.com/20375610/192454561-65377837-268f-4d28-9a30-f991c056c5bb.png)

Removing `<API, RegisteredSubs>` for `link` means `API` gets default to `unknown` as you've added `API extends Web3APISpec = unknown,` which yields:

![image](https://user-images.githubusercontent.com/20375610/192457431-47f4d380-8869-4a2b-a7aa-432fa84730db.png)

which means TypeScript is no longer aware of what's available for the `requestManager` to call, what `params` the specified `method` takes, and what the return type of the request is

Reply:
Actually, `requestManager` and `link<T extends Web3Context>` are separate. And so not passing `API` to `link<T extends Web3Context>` does not affect `requestManager`. And I kept it untouched as-is:
```typescript
export class Web3RequestManager<
	API extends Web3APISpec = EthExecutionAPI,
> extends Web3EventEmitter<{
	[key in Web3RequestManagerEvent]: SupportedProviders<API> | undefined;
}> {
...
```
And so the safe typing is working well:
![image](https://user-images.githubusercontent.com/24407834/192628813-0bf87276-3dd4-4952-859b-e1ef31adb13a.png)


And, if I provided no `params` to `eth_getBalance` then running `yarn build` would give the following error:
> web3-eth: src/rpc_methods.ts(99,3): error TS2322: Type '[]' is not assignable to type '[address: string, blockNumber: BlockNumberOrTag]'.

So, I think you just had a cashing problem when checking out this branch. Kindly try again by running `yarn build` close vscode and reopen it, if it was still not showing you the correct error...

Reply:
The third screenshot is the inferred type when using the `requestManager` for the plugin. The `EthExecutionApi` (i.e. the `requestManager` when used within the `web3-eth` package) still resolves as expected with your approach, but the same `API` type resolution does not occur for `ChainlinkPluginAPI` - this is what #5480 is attempting to solve. I don't think making use of `unknown` is the best solution for this issue as we do know what the `API` is because we [specify it](https://github.com/web3/web3.js/blob/3e3cb8e20e7eb9909f4082faa0487b3d54e67cb2/packages/web3-plugin-example/src/index.ts#L34). The changes in #5480 should allow us to pass an `API` that's not `EthExecutionApi` (or extends it as `Web3EthExecutionApi` does), but as mentioned on our call and below, our code around subscriptions/logs is hardcoded to handle `EthExecutionApi` and therefore throws type errors when trying to use an `API` that doesn't have `eth_subscribe`

Reply:
Yea I was totally thinking about `ChainlinkPluginAPI` incorrectly - it's not needed given that there are no JSON RPC methods being added as you've mentioned. I think it's fine for `Web3PluginBase` to just default to `Web3APISpec` (so the user doesn't have to worry about passing the generic if they're not adding a JSON RPC API) as added by [this](https://github.com/web3/web3.js/pull/5393/commits/816f2f58ab21113ee4dfea32cc415813880d3381) commit

The rest of the changes you've made in this PR (namely around `Web3Context`) I don't think are needed for #5393, but probably do need to be incorporated into #5480. #5480 is actually trying to solve an issue not strictly related to #5393 and that's fixing the existing architecture around `Web3Context` and the `RequestManager` when you pass an `API` that doesn't extend `EthExecutionApi` - which I don't believe should be a requirement when a plugin/user wants to add JSON RPC methods (it's currently a requirement since code around log subscriptions is currently hardcoded to use `EthExecutionApi` as mentioned [here](https://github.com/web3/web3.js/pull/5416#issuecomment-1259084790))

Reply:
Great @spacesailor24 :+1: 
I am happy to see that this MR helped resolve the 2 blockers for https://github.com/web3/web3.js/pull/5393 and provided clarifications :smile: 

And so it is time to merge this MR. 

However, because I think that, there are some copied ideas from this MR to  https://github.com/web3/web3.js/pull/5393 (and this is also the reason for having conflicts in the MR with the destination branch). I suggest you do one of the following:

- Ideally, it would be better to preserve the history and the orders of the commits. And so I suggest you to `rebase` your branch to an earlier commit and then re-push the commits from this branch and from your branch in the order they came. And before doing this I recommend that you save the commits from both branches and the date of each one. And also to have a full copy of the local folder, that you can use in case things were messed-up. This will keep the contribution clean of who did what at which time and it would be easier to trace if later needed and this is the best practice, I think.
- The other option is easier but will vanish the valuable history. It is to resolve the conflicts in this MR and merge it with yours. This is not the best option because there were some ideas copied from this MR to the other MR in parallel instead of merging it early. But I am OK with that if you faced issues when trying the first option.

And by the way, I also pushed some more enhancements and comments to the code so when someone later would implement a plugin would have a clearer understanding, I hope. And this included 2 classes `Web3PluginBase` and  `Web3EthPluginBase` :smile:.